### ISSUE TITLE
Object property can't be set as nullable: true

### ISSUE THREAD

**Issue by bajdka** [domaindrivendev/Swashbuckle.AspNetCore#1589]
**Used versions:**
Swashbuckle.AspNetCore 5.2.0
Swashbuckle.AspNetCore.Newtonsoft 5.2.0

**Steps to reproduce:**
I created response model with object property, expecting `offer` to be nullable. I tried two different approaches, without results:
- allowing null property
```
    public class DashboardGetResponse
    {
        [JsonProperty(Required =  Required.AllowNull)]
        public OfferGetModel Offer { get; set; }
    }
```
- allowing null directly inside object object
```
    [JsonObject(ItemRequired =  Required.AllowNull)]
    public class OfferGetModel
    {
        public Guid OfferId { get; set; }
    }
```

**Actual result:**
- neither `DashboardGetResponse` nor `OfferGetModel` schema suggests that `offer` field is nullable
```
"OfferGetModel": {
        "type": "object",
        "properties": {
        "offerId": {
            "type": "string",
            "format": "uuid"
          }
        }
 }
```

```
 "DashboardGetResponse": {
        "required": [
          "offer"
        ],
        "type": "object",
        "properties": {
          "offer": {
            "$ref": "#/components/schemas/OfferGetModel"
          }
     }
}
```
**Expected result:**
- `nullable: true` property in `OfferGetModel` swagger schema  
```
"OfferGetModel": {
        "type": "object",
        "properties": {
        "offerId": {
            "type": "string",
            "format": "uuid"
          },
        },
       "nullable": true
      },
```
OR
- `DashboardGetResponse` referencing that `offer` object can be of nullable type

**Comment by domaindrivendev** [comment]
This is a complicated one to explain, particularly if you're not familiar with the [Swagger/OpenAPI Spec](https://swagger.io/docs/specification/about/), but I'll give it a go.

By design, object properties are created as "reference" schema's rather than "inline" schemas. This means that the property schema will hold a reference to a shared schema that's defined in the `components` section of the generated document. Now, if that "shared" definition had the `nullable` property set, then that would mean it applies to all uses of the schema. In most cases this would be OK but not all. For example, if the schema was used as a response model, then you're documentation will state that the response from an operation MAY be null, and this could be very misleading.

Long story short, the `nullable` flag (as well as some other schema properties) are "contextual" and therefore applying them to a "shared" schema definition ends up being very problematic. So, out-of-the-box, SB just doesn't do it.

In fact, this is a [known issue/consideration](https://github.com/OAI/OpenAPI-Specification/issues/1368) when working with Swagger/Open API documents. The recommended approach is to create "inline" schemas that "extend" the shared schema by using the `allOf` keyword. Swashbuckle DOES support this approach but it's opt-in. That is, if you apply the `UseAllOfToExtendReferenceSchemas` setting, it will structure the generated Swagger slightly differently, in a way that _should_ resolve your issue.

**Comment by samiroquai** [comment]
I've read the discussion on the OpenAPI repository (https://github.com/OAI/OpenAPI-Specification/issues/1368). In fact it led me here. Good summary. I've tried your suggestion and it works :) Thank you.
